### DISCUSSION TITLE
LaTeX in PyVista Plots

### DISCUSSION THREAD

**Discussion by dominik-kern** [pyvista/pyvista#2928] [Q&A] [answered]
Is there any chance to use a TeX interpreter on labels such as scalar_bar?
This would enable a uniform appearance and improve the integration into documents, right now I reduce all labels to the bare minimum and write about them in the text, anyway for greek symbols, small expression this would be very useful.
In the documentation I found only how to change the font.

**Comment by guido-granda** [comment]
Hello, I also have the same problem. But, I know that Latex used to work in previous version of pyvista. Look at this picture 
![save_df_false_raster](https://user-images.githubusercontent.com/7574434/200379121-9324e215-e105-4bba-aff3-94252479c0bb.svg)

**Comment by banesullivan** [comment]
This requires VTK to be built with mathtext support. Are you installing from PyPI or from conda-forge? I think the wheels on PyPI are built with this by default (but I'm not 100% sure)

I don't think this is enabled for the conda-forge VTK build, see https://github.com/conda-forge/vtk-feedstock/issues/79

A quick way to tell if this is supported in your build of VTK is the following:

```py
>>> import vtk
>>> vtk_mathtext = vtk.vtkMathTextFreeTypeTextRenderer()
>>> vtk_mathtext.MathTextIsSupported()
True
```

**Reply by guido-granda** [reply]
Hi, I ran these three lines and I also got True. I also did a conda list of my environment and I got that pyvista 0.37.0 and vtk 9.2.2 where installed using the pypi channel. So from what you said mathtext support is enable and since it is not working there has to be another reason?

**Reply by banesullivan** [reply]
Hm I'm noticing this too. We may need to find a previous version PyVista where this was working and do a git bisect.

**Reply by guido-granda** [reply]
Hello, 
I did a conda list revisions and obtained the following about vtk and pyvista when latex support was working:

vtk  {8.2.0 (conda-forge/linux-64) -> 9.0.3 (defaults/linux-64)}
pyvista  {0.32.1 (conda-forge/noarch) -> 0.34.0 (conda-forge/noarch)}

**Reply by banesullivan** [reply]
Following up. This broke within VTK for matplotlib 3.6. This will/should be fixed in VTK 9.2.3 (handled in [!9750](https://gitlab.kitware.com/vtk/vtk/-/merge_requests/9750))

I got it to work locally by downgrading matplotlib then re-installing the latest VTK (reinstalling VTK seemed necessary for me to get this to work)

```py
import pyvista as pv
p = pv.Plotter()
p.add_text(r'$\rho$')
p.show()
```

<img width="624" alt="Screen Shot 2022-12-13 at 1 54 57 PM" src="https://user-images.githubusercontent.com/22067021/207441631-3fc40b04-af30-4130-a4bf-d82c2fe45f25.png">

**Reply by zhangruoyao68** [reply]
Hi,

Which version of matplotlib is working with the latest VTK? I am also having trouble with matplotlib 3.6.2, pyvista 0.37.0 and vtk 9.2.4. Thanks!

**Reply by banesullivan** [reply]
matplotlib 3.6.2 and vtk 9.2.4 *should* work for this... I just checked locally and all works for me on my Arm-based Mac. Specifially, this is my environment:

```py
>>> import pyvista as pv
>>> pv.Report()

--------------------------------------------------------------------------------
  Date: Thu Jan 05 19:42:28 2023 MST

                OS : Darwin
            CPU(s) : 8
           Machine : arm64
      Architecture : 64bit
               RAM : 24.0 GiB
       Environment : Python
       File system : apfs
        GPU Vendor : Apple
      GPU Renderer : Apple M2
       GPU Version : 4.1 Metal - 76.3

  Python 3.10.8 | packaged by conda-forge | (main, Nov 22 2022, 08:25:29)
  [Clang 14.0.6 ]

           pyvista : 0.38.dev0
               vtk : 9.2.4
             numpy : 1.23.5
           imageio : 2.22.4
            scooby : 0.7.0
             pooch : v1.6.0
        matplotlib : 3.6.2
           IPython : 8.7.0
          colorcet : 3.0.1
           cmocean : 2.0
        ipyvtklink : 0.2.3
              tqdm : 4.64.1
            meshio : 5.3.4
         pythreejs : Version unknown
    pytest_pyvista : 0.1.1
--------------------------------------------------------------------------------
```

**Reply by zhangruoyao68** [reply]
This is interesting. My environment is
```python
>>> import pyvista as pv
>>> pv.Report()

--------------------------------------------------------------------------------
  Date: Thu Jan 05 21:46:12 2023 EST

                OS : Darwin
            CPU(s) : 10
           Machine : arm64
      Architecture : 64bit
               RAM : 32.0 GiB
       Environment : Python
       File system : apfs
        GPU Vendor : Apple
      GPU Renderer : Apple M1 Max
       GPU Version : 4.1 Metal - 83

  Python 3.9.15 (main, Nov 24 2022, 08:28:41)  [Clang 14.0.6 ]

           pyvista : 0.37.0
               vtk : 9.2.4
             numpy : 1.21.5
           imageio : 2.19.3
            scooby : 0.7.0
             pooch : v1.6.0
        matplotlib : 3.6.2
             PyQt5 : 5.15.7
           IPython : 8.7.0
          colorcet : 3.0.1
             scipy : 1.9.3
              tqdm : 4.64.1
        jupyterlab : 3.5.2
--------------------------------------------------------------------------------
```

the outcome of running your example is

<img width="838" alt="Screenshot 2023-01-05 at 9 58 17 PM" src="https://user-images.githubusercontent.com/25314650/210920901-7071dc11-b369-4b99-b5da-6591e2bbe13e.png">

**Reply by banesullivan** [reply]
Well, that is quite odd. I wonder if the order in which you installed these packages matters. Would you be up for creating a new, fresh virtual environment, installing pyvista, and trying again?

**Reply by banesullivan** [reply]
FYI, I added a regression test for this in https://github.com/pyvista/pyvista/pull/3781 which may help us uncover which environments and combinations of VTK+matplotlib support this

**Reply by kaathaarinaa** [reply]
Hi, I continue to have the same issue. I tried the following combinations (none of them work):
matplotlib 3.6.2 vtk 9.2.2 pyvista 0.37.0
matplotlib 3.6.2 vtk 9.2.2 pyvista #3910 (some branch towards 0.38.0)
matplotlib 3.6.2 vtk 9.2.5 pyvista #3910 (some branch towards 0.38.0)
matplotlib 3.6.2 vtk 9.2.4 pyvista #3910 (some branch towards 0.38.0)

and all of them with reinstalling vtk after pyvista.

I ran both the tests suggested above:
```
>>> import vtk
>>> vtk_mathtext = vtk.vtkMathTextFreeTypeTextRenderer()
>>> vtk_mathtext.MathTextIsSupported()
True
```
and
```
import pyvista as pv
p = pv.Plotter()
p.add_text(r'$\rho$')
p.show()
```
but the result is just a text output of `$\rho$` not $\rho$...

Any update on this issue?

**Reply by banesullivan** [reply]
For this specifically, your version of PyVista will not matter too much.

For Matplotlib 3.6.2, only VTK 9.2.5 is compatible.

For VTK<9.2.5, you must use Matplotlib<3.6

----

How are you installing VTK? With `pip` or `conda` or a custom build?

Can you share a `pv.Report()`?

Can you try re-installing both matplotlib and VTK when you have compatible versions?

I recently added good image regression tests for this in #3781 and further clarified the compatible versions in #3838. I feel pretty confident that you just need to have clean installations of compatible VTK and Matplotlib versions

**Reply by kaathaarinaa** [reply]
ok, so I install vtk using ```pip```.
Therefore, after uninstalling matplotlib and vtk, as you suggested, I do:
```
pip install matplotlib==3.6.2
pip install vtk==9.2.5
```
the result of the two tests after reinstall is:
```
# Test 1
>>> import vtk
>>> vtk_mathtext = vtk.vtkMathTextFreeTypeTextRenderer()
>>> vtk_mathtext.MathTextIsSupported()
True
# So that should be fine
```
```
# Test 2
import pyvista as pv
p = pv.Plotter()
p.add_text(r'$\rho$')
p.show()
# not working and still displaying $\rho$
```

```pv.Report()```:
```
  Date: Tue Jan 31 15:42:45 2023 Mitteleuropäische Zeit

                OS : Windows
            CPU(s) : 48
           Machine : AMD64
      Architecture : 64bit
               RAM : 63.9 GiB
       Environment : Python
        GPU Vendor : ATI Technologies Inc.
      GPU Renderer : AMD RADEON HD 6450
       GPU Version : 4.5.13417 Core Profile Context 0

  Python 3.8.2 (tags/v3.8.2:7b3ab59, Feb 25 2020, 23:03:10) [MSC v.1916 64 bit
  (AMD64)]

           pyvista : 0.38.dev0
               vtk : 9.2.5
             numpy : 1.22.4
           imageio : 2.22.4
            scooby : 0.7.1
             pooch : v1.6.0
        matplotlib : 3.6.2
           IPython : 8.7.0
             scipy : 1.9.3
              tqdm : 4.64.1
```

**Reply by banesullivan** [reply]
Oh hm. Windows is always a wildcard. I'll see if I can break out my windows machine and see if I can get it working there.

This line worries me that it may not be enabled:

https://gitlab.kitware.com/vtk/vtk/-/blob/master/.gitlab/ci/configure_windows.cmake#L20

**Reply by kaathaarinaa** [reply]
Thanks for the reply. I am not sure I fully understand the relevance, but I could go for an editable install of vtk and just change that line if that helps?

**Reply by banesullivan** [reply]
I recently confirmed on a windows machine that math text symbols are not working... This seems like an upstream VTK issue to me.

**Comment by Joalland** [comment]
Hi there. This may be strange but when I do :
```python
import pyvista as pv
p = pv.Plotter()
p.add_text(r'$\rho$')
p.show()
# not working and still displaying $\rho$
```
it doesn't produce the latex.

But then i open another shell and I do :
```python
import vtk
import pyvista as pv
p = pv.Plotter()
p.add_text(r'$\rho$')
p.show()
#it works !
```
it works. (However it doesn't work if I don't open another shell).

My pv.Report() is :
```
--------------------------------------------------------------------------------
  Date: Tue Mar 07 18:58:29 2023 +04

                OS : Linux
            CPU(s) : 8
           Machine : x86_64
      Architecture : 64bit
               RAM : 31.2 GiB
       Environment : IPython
       File system : ext4
        GPU Vendor : Intel
      GPU Renderer : Mesa Intel(R) UHD Graphics (CML GT2)
       GPU Version : 4.6 (Core Profile) Mesa 21.2.6

  Python 3.8.10 (default, Nov 14 2022, 12:59:47)  [GCC 9.4.0]

           pyvista : 0.37.0
               vtk : 9.2.2
             numpy : 1.22.3
           imageio : 2.21.2
            scooby : 0.7.1
             pooch : v1.6.0
        matplotlib : 3.5.3
             PyQt5 : 5.14.1
           IPython : 7.13.0
          colorcet : 3.0.1
             scipy : 1.7.3
              tqdm : 4.64.1
            meshio : 5.3.4
        jupyterlab : 3.6.1
--------------------------------------------------------------------------------
```

**Reply by kaathaarinaa** [reply]
I had to test this, so here are my results:
with vtk 9.0.3 and pyvista 0.38.3 --> 
```
Fatal Python error: Cannot recover from stack overflow.
Python runtime state: initialized
.
.
.

Process finished with exit code -1073740791 (0xC0000409)
```
BUT =D:
with vtk 9.2.6 and pyvista 0.38.3 --> same result as @Joalland 
so it produces proper Tex output only if `import vtk` is used...
Here is my pv.Report():
```
--------------------------------------------------------------------------------
  Date: Tue Mar 07 16:11:30 2023 Mitteleuropäische Zeit

                OS : Windows
            CPU(s) : 4
           Machine : AMD64
      Architecture : 64bit
               RAM : 31.9 GiB
       Environment : Python
        GPU Vendor : NVIDIA Corporation
      GPU Renderer : GeForce GT 1030/PCIe/SSE2
       GPU Version : 4.5.0 NVIDIA 456.71

  Python 3.8.2 (tags/v3.8.2:7b3ab59, Feb 25 2020, 23:03:10) [MSC v.1916 64 bit
  (AMD64)]

           pyvista : 0.38.3
               vtk : 9.2.6
             numpy : 1.20.3
           imageio : 2.9.0
            scooby : 0.7.1
             pooch : v1.6.0
        matplotlib : 3.4.3
             PyQt5 : 5.12.3
           IPython : 7.28.0
             scipy : 1.7.1
              tqdm : 4.62.2
      nest_asyncio : 1.5.1
--------------------------------------------------------------------------------
```

**Reply by banesullivan** [reply]
I think you need to note my comment above: https://github.com/pyvista/pyvista/discussions/2928#discussioncomment-4830189

> For this specifically, your version of PyVista will not matter too much.
>
> For Matplotlib 3.6.2, only VTK 9.2.5+ is compatible.
>
> For VTK<9.2.5, you must use Matplotlib<3.6

**Reply by guido-granda** [reply]
Hi, good to know there is a way to solve this issue. How can I know which vtk and matplotlib versions are compatible and produce good latex text?

**Reply by akaszynski** [reply]
We might want to consider introducing a runtime check for this.

**Reply by banesullivan** [reply]
It would be effectively the same as the check in our CI: https://github.com/pyvista/pyvista/blob/c4cbfdb526a908d6282e4feb202d1f6f266c18ed/tests/plotting/test_plotting.py#L2618-L2619

**Reply by guido-granda** [reply]
Hi, I just realize that with this solution of import vtk it is not possible to use loops. The code just crash saying kernel crashed :-(

**Reply by LyceanEM** [reply]
Has there been any update to this issue? I have tried changing my vtk and matplotlib versions as discussed, but it is causing kernel crashes, and error messages like 
```
PyVista error in handling VTK error message:
'WARNING:root:WARNING:root:WARNING:root:WARNING:root:WARNING:root:WARNING:root:WARNING:root:WARNING:r' ... (3791 characters)
```

**Reply by benaplus1** [reply]
I've been having the same issue. Trying to plot several times of a 3D model field, and vtk throws a fit when using a for loop to iterate over the times.

**Reply by banesullivan** [reply]
As far as I am aware, this is resolved in PyVista and my comment in an above thread is the most critical part: https://github.com/pyvista/pyvista/discussions/2928#discussioncomment-4830189

> For this specifically, your version of PyVista will not matter too much.
>
> For Matplotlib 3.6.2, only VTK 9.2.5+ is compatible.
>
> For VTK<9.2.5, you must use Matplotlib<3.6

I've confirmed this is still the case with PyVista on both my Mac and Linux systems with the latest versions of VTK and PyVista. 

----

Considering there are several of you still experiencing issues, I'm considering opening a new issue to get to the bottom of what specifically is going wrong or what common thread there may be with you all (I suspect you all may be Windows users?). 

Please open a new issue where you share a `pv.Report()` (which FYI should tell you if MathText is supported in your environment)

**Reply by FrankFrank9** [reply]
Hello, 
I'm facing this issue again even with the latest version of pyvsta.

The report is:

`                
             OS : Linux
            CPU(s) : 32
           Machine : x86_64
      Architecture : 64bit
               RAM : 251.5 GiB
       Environment : Python
       File system : xfs
        GPU Vendor : Mesa/X.org
      GPU Renderer : llvmpipe (LLVM 15.0.6, 256 bits)
       GPU Version : 4.5 (Core Profile) Mesa 22.3.6
  MathText Support : False

    Python 3.11.0 (main, Mar  1 2023, 18:26:19) [GCC 11.2.0]
  
             pyvista : 0.43.0
                 vtk : 9.3.0
               numpy : 1.26.2
          matplotlib : 3.8.2
              scooby : 0.9.2
               pooch : 1.8.0
              pillow : 10.1.0
             imageio : 2.33.0
               PyQt5 : 5.15.10
             IPython : 8.15.0
               scipy : 1.11.4
                tqdm : 4.66.1
        nest_asyncio : 1.5.6
  
    Intel(R) oneAPI Math Kernel Library Version 2023.1-Product Build 20230303
    for Intel(R) 64 architecture applications
`

But if run 

`>>> import vtk
>>> vtk_mathtext = vtk.vtkMathTextFreeTypeTextRenderer()
>>> vtk_mathtext.MathTextIsSupported()
True`

and also vtk and matplotlib seem to have the correct version. 

The only thing is that I'm running in headless mode, can this have any effect?
Moreover, locally I have a GPU which is not detected and I'm still running with Mesa instead of EGL? Any advice on that as well?

Best regards

**Reply by banesullivan** [reply]
@FrankFrank9, this is odd. Because `pv.Report()` is showing `MathText Support : False` for you. Are you sure `vtk.vtkMathTextFreeTypeTextRenderer().MathTextIsSupported()` is `True`?

**Reply by FrankFrank9** [reply]
@banesullivan Yes, 100% sure, I do agree with you that it is very odd but I do not have an explanation for that...

**Reply by Steriva** [reply]
I have faced a similar issue recently. I don't know if this problem has already been solved, nevertheless I will post it for completeness.

At first, I performed the following test
```python
import vtk
vtk_mathtext = vtk.vtkMathTextFreeTypeTextRenderer()
vtk_mathtext.MathTextIsSupported()
```
getting _True_ as expected. Then, if I try to use add a Latex expression in PyVista with 
```python
import pyvista as pv
plotter = pv.Plotter()
plotter.add_text(r'$\rho$')
plotter.show()
```
The code outputs _$\rho$_, so I have done the following
```python
import vtk
import pyvista as pv
plotter = pv.Plotter()
plotter.add_text(r'$\rho$')
plotter.show()
```
and the _rho_ is displayed correctly.

Best,
SR

**Reply by guido-granda** [reply]
@Steriva can you do a pv.Report() and post the result here please

**Reply by Steriva** [reply]
Yes, of course.
Without the vtk import
```python
import pyvista as pv
pv.Report()
```
i get the following
```
OS	Linux	CPU(s)	16	Machine	x86_64
Architecture	64bit	RAM	31.1 GiB	Environment	Jupyter
File system	ext4	GPU Details	error	MathText Support	False
Python 3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:40:32) [GCC 12.3.0]
pyvista	0.42.2	vtk	9.2.6	numpy	1.23.5
matplotlib	3.8.0	scooby	0.7.2	pooch	v1.7.0
pillow	10.0.0	PyQt5	5.15.9	IPython	8.15.0
ipywidgets	7.7.2	scipy	1.11.2	tqdm	4.66.1
meshio	5.3.4	jupyterlab	4.0.6	nest_asyncio	1.5.8
```
on the other hand importing vtk
```python
import vtk
import pyvista as pv
pv.Report()
```
the output is 
```
OS	Linux	CPU(s)	16	Machine	x86_64
Architecture	64bit	RAM	31.1 GiB	Environment	Jupyter
File system	ext4	GPU Details	error	MathText Support	True
Python 3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:40:32) [GCC 12.3.0]
pyvista	0.42.2	vtk	9.2.6	numpy	1.23.5
matplotlib	3.8.0	scooby	0.7.2	pooch	v1.7.0
pillow	10.0.0	PyQt5	5.15.9	IPython	8.15.0
ipywidgets	7.7.2	scipy	1.11.2	tqdm	4.66.1
meshio	5.3.4	jupyterlab	4.0.6	nest_asyncio	1.5.8
```

**Reply by banesullivan** [reply]
We should start a thread on the VTK discourse about this as it seems VTK and `vtk.vtkMathTextFreeTypeTextRenderer().MathTextIsSupported()` has inconsistent behavior 

https://discourse.vtk.org/

**Reply by henry2004y** [reply]
I am also using windows but the rendered latex code snippet above does not work for me.

**Reply by user27182** [reply]
> We should start a thread on the VTK discourse about this as it seems VTK and `vtk.vtkMathTextFreeTypeTextRenderer().MathTextIsSupported()` has inconsistent behavior 
> 
> 
> 
> https://discourse.vtk.org/

There are magic vtk imports needed to ensure this check returns the correct result 

```
>>> import vtkmodules.vtkRenderingFreeType
>>> import vtkmodules.vtkRenderingMatplotlib
>>> vtkmodules.vtkRenderingFreeType.vtkMathTextFreeTypeTextRenderer().MathTextIsSupported()
True
```

See https://discourse.vtk.org/t/how-to-check-if-mathtext-is-supported-without-importing-all-of-vtk/16038

**Reply by user27182** [reply]
> > We should start a thread on the VTK discourse about this as it seems VTK and `vtk.vtkMathTextFreeTypeTextRenderer().MathTextIsSupported()` has inconsistent behavior
> > https://discourse.vtk.org/
> 
> There are magic vtk imports needed to ensure this check returns the correct result
> 
> ```
> >>> import vtkmodules.vtkRenderingFreeType
> >>> import vtkmodules.vtkRenderingMatplotlib
> >>> vtkmodules.vtkRenderingFreeType.vtkMathTextFreeTypeTextRenderer().MathTextIsSupported()
> True
> ```
> 
> See https://discourse.vtk.org/t/how-to-check-if-mathtext-is-supported-without-importing-all-of-vtk/16038

These imports are added in https://github.com/pyvista/pyvista/pull/8086 and should finally be fixed in the next release, `v0.47`.

In the interim, users should include these two imports to make LaTeX rendering work:

``` py
import vtkmodules.vtkRenderingFreeType
import vtkmodules.vtkRenderingMatplotlib
import pyvista as pv
p = pv.Plotter()
p.add_text(r'$\rho$')
p.show()
```

**Comment by user27182** [comment, marked as answer]
As mentioned here https://github.com/pyvista/pyvista/discussions/2928#discussioncomment-14938454, the fix is to do:

``` py
import vtkmodules.vtkRenderingFreeType
import vtkmodules.vtkRenderingMatplotlib
import pyvista as pv
p = pv.Plotter()
p.add_text(r'$\rho$')
p.show()
```
or upgrade to pyvista v0.47 (not yet released).

I'm marking this as the answer. Please comment if LaTeX rendering is still not working for you.
